docs(accessibility-testing): Updates content as part of content audit. (#3776)
patternfly/patternfly-org: packages/documentation-site/patternfly-docs/content/developer-resources/accessibility-testing.md

  - [Voiceover](https://support.apple.com/guide/voiceover/welcome/mac) with Safari, Mac ([Voiceover keyboard shortcuts](//dequeuniversity.com/screenreaders/voiceover-keyboard-shortcuts))
  - [NVDA](https://www.nvaccess.org/download/) with Firefox, Windows ([NVDA keyboard shortcuts](//dequeuniversity.com/screenreaders/nvda-keyboard-shortcuts))

### Check color contrast

Colors should pass the following contrast checks, to ensure that users across the vision spectrum can understand your product:
    - Text color against background color ([Understanding WCAG 1.4.3](//www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html))
    - Text color against link color ([Technique G183](//www.w3.org/TR/WCAG20-TECHS/G183.html))
    - Visible boundaries of buttons and form elements against adjacent background color ([Understanding WCAG 1.4.11](//www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html))

## Accessibility testing checklist

We recommend referencing the following checklist as you assess your product's accessibility.

> These are some of the main areas we check for on the PatternFly team to ensure that components meet consistent accessibility standards, but, to check your complete implementation, we recommend checking these same areas in your products. 

Author: ```suggestion
These are some of the main areas we check for on the PatternFly team to ensure that a UI meets consistent accessibility standards. To assess your specific implementation, we recommend checking these same areas in your product. 
```

<Checkbox label="Heading levels convey structure/content and do not skip levels." description="A common practice is to use a single h1 for the primary headline or logo on a page, h2s for designating major sections, and h3s for supporting sections." id="accessibility-points-7b" />
<Checkbox label="Landmarks" id="accessibility-points-7c" />
<Checkbox label="Links" id="accessibility-points-7d" />
<Checkbox label="Form controls" id="accessibility-points-7e" />
<Checkbox label="Tables" id="accessibility-points-7f" /></>}  id="accessibility-points-7" />
<br/>
<Checkbox label={<span> <b> Labels: </b> </span>} body={<><Checkbox label=<span><b>Link</b> labels  are descriptive, informative, and unique (unless they have the same URL).</span> id="accessibility-points-8a"/> 

  <Checkbox label={<span><b>Buttons and form controls:</b></span>} description={ <span>
    <li> All form controls have clear and descriptive labels. </li>

>     <li> Indicate that a button has the expandable control. </li>

Reviewer: I think there should be some clarification on this one that this line is only applicable to expandable buttons, not all buttons.

Reviewer 2: Could be worth explicitly mentioning `aria-expanded` and how it should always have a boolean value if the button is intended to be expandable

  </span>
  }
  id="general-criteria-8b" />
    
    <Checkbox label={<span><b>Form inputs</b> have a label (even if it's not visible).</span>} id="general-criteria-8c" />
    
    <Checkbox label={<span><b>Icons</b> have some some kind of text for screen readers (even if it's not visible.)</span>} id="general-criteria-8d" />

>     <Checkbox label={<span><b>Images have proper alt text.</b></span>} description="The exception to this practice is when images are primarily for presentation purposes and are not essential pieces of content. To signify that an image should be skipped by a screen reader, set the value of the alt attribute to an empty string: `alt=””`." id="general-criteria-8e" />

Reviewer 2: We just need to replace the backticks with the manual ws-code span:

```suggestion
    <Checkbox label={<span><b>Images have proper alt text.</b></span>} description={<span>The exception to this practice is when images are primarily for presentation purposes and are not essential pieces of content. To signify that an image should be skipped by a screen reader, set the value of the alt attribute to an empty string: <code className="ws-code">alt=""</code></span>}  id="general-criteria-8e" />
```

    <Checkbox label={<span> <b>Landmark</b> regions should have labels when there is more than one type of landmark, and they are not the same (such as navigation, main, form, etc.). Section element shouldn’t be used unless it has a label.</span>} id="general-criteria-8f" />

>     <Checkbox label={<span><b>Tables</b> and table content is clearly described. <a href="https://webaim.org/techniques/tables/">WebAIM has additional guidance on this.</a></span>} id="general-criteria-8g" />

Reviewer 2: Just to tweak the link text, maybe something like:

```suggestion
    <Checkbox label={<span><b>Tables</b> and table content is clearly described. WebAIM has additional guidance on <a href="https://webaim.org/techniques/tables/">creating accessible tables</a>.</span>} id="general-criteria-8g" />
```

>     <Checkbox label={<span><b>Aria</b> provides descriptive details to screen reader users, without reiterating or overriding the text that is already there. If there is visible text then there doesn’t need to be an <span className="ws-code">aria-label</span>. </span>} description={<span>It’s important to understand that ARIA can only affect the semantics of an element; not the behavior. For example, while you can hide an element from screen readers with <span> className="ws-code">aria-hidden=”true”</span>, that does not change the focus behavior for that element.</span>} id="general-criteria-8h" />

Reviewer 2: @jessiehuff wdyt about tweaking this verbiage? Feel like we could name `aria-label` as a way to add a non-visual label to certain elements, and `aria-labelledby` to use an existing text or aria-label to label another element. 

Though if this is meant to be more of a general point about aria rather than strictly labeling, maybe the verbiage about when you don't need an aria-label could be its own bullet?

Additionally, looks like the description has a typo in the description. `<span> className="ws-code">aria-hidden=”true”</span>` should be `<span className="ws-code">aria-hidden=”true”</span>`

Reviewer: Yeah, it was originally intended as more of a general statement of when something needs to be labeled vs. not, but admittedly I probably shouldn't have just singled out `aria-label` in that circumstance. (I started calling this out because I noticed that once people started trying to improve their a11y, they could at times just go crazy adding `aria-label`s even when there was no need for one and adding one might even make the experience worse.) Maybe we could change the `aria-label` in this context to a label in general and then call out your point in a bullet.

    <Checkbox label={<span> <b>Landmark</b> regions should have labels when there is more than one type of landmark, and they are not the same (such as navigation, main, form, etc.). Section element shouldn’t be used unless it has a label.</span>} id="general-criteria-8f" />

    <Checkbox label={<span><b>Tables</b> and table content is clearly described. WebAIM has additional guidance on <a href="https://webaim.org/techniques/tables/">creating accessible tables</a>.</span>} id="general-criteria-8g" />

    <Checkbox label={<span><b>ARIA</b> labels provide descriptive details to screen reader users, without reiterating or overriding the text that is already there. If there is visible text then there doesn’t need to be an ARIA label. </span>} 
    description={ <span>
      <li> To add a non-visual label to an element you should use <span className="ws-code">aria-label</span>. Keep in mind you should not add this to an element that already has a visual label in the UI. You can also use <span className="ws-code">aria-labelledby</span> to label an element through existing text or <span className="ws-code">aria-label</span>. </li>

>       <li> <span> It’s important to understand that ARIA can only affect the semantics of an element; not the behavior. For example, while you can hide an element from screen readers with <span className="ws-code">aria-hidden=”true”</span>, that does not change the focus behavior for that element.</span></li>

Reviewer 2: Not sure if this really fits here anymore since it's more about labels than general ARIA. @jessiehuff wdyt about whether this should stay here, removed, moved to the Product Dev Guide page, or a secret 4th option that I haven't found yet?

Reviewer: Yeah, I see your point @thatblindgeye. Maybe it's best to just remove those subpoints? It looks like the the point about aria only affecting the semantics is already in the dev guide page, but do we have a section where we clarify the difference between `aria-label` and `aria-labelledby`? I wonder if it would make sense to remove that point here but just make sure that we have it clear in the dev docs somewhere?

Author: @thatblindgeye @jessiehuff I can see how the aria-labelledby point may be more helpful/visible in the dev guide. I went ahead and removed the subpoints here - I can add the aria-labelledby in a separate issue (or slip it into this PR if that seems okay) - wdyt?

section: accessibility
---

# Testing your product's accessibility 

> The following guide contains instructions and recommendations to comprehensively test your product's accessibility to identify accessibility issues and opportunities for improvement.  

Author: ```suggestion
This guide contains instructions and recommendations to comprehensively test your product's accessibility to identify accessibility issues and opportunities for improvement.  

**Keep in mind that this guide may not cover every scenario.**

```

## General testing recommendations 

Many accessibility issues can be found by doing a few simple checks: 

> ### 1. Validate your HTML

Author: Yeah - I'll update that! I was 50/50 on doing this & mostly wanted to visualize what it'd look like, but since you mentioned it I think that's good confirmation to just keep normal headings ty 😊

> ### 1. Validate your HTML

Author: ```suggestion
### Validate your HTML
```

Good accessibility practices start with structural, semantic HTML. When a screen reader (or any sort of assistive device) scans a web page, it gets information about the Document Object Model (DOM), or the HTML structure of the page. No styles or JavaScript will be read by a screen reader. 

Screen reader software like Voice Over (VO), NVDA, or JAWS doesn’t just turn text into speech. It can use information in the HTML to list all of the headings on a page, give extra navigation controls to data tables, announce how many items are in a list, and more. This makes semantic HTML essential. 

There are many tools you can use to validate your HTML, such as [W3C’s markup validation service](https://validator.w3.org/). 

> ### 2. Check for accessibility violations with an audit tool

Author: ```suggestion
### Check for accessibility violations with an audit tool
```

>  If you are using PatternFly, we recommend using [aXe: The Accessibility Engine](//www.deque.com/axe/) to check for accessibility violations locally. To make it even easier for you, we've created the [patternfly-a11y script](https://github.com/patternfly/patternfly-a11y), which reports any aXe accessibility violations from a set of pages. 
>  
>  To use this script, a configuration file can be set to adapt our script to your specific needs like authentication, waiting for any specific items to finish loading (like a loading spinner), etc. 

Author: ```suggestion
 If you are using PatternFly, we recommend using aXe: The Accessibility Engine (using their [DevTools](https://www.deque.com/axe/devtools/) or the [Chrome](https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd) or [Firefox](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/) extension) to check for accessibility violations locally. 
 
We also offer the [patternfly-a11y script](https://github.com/patternfly/patternfly-a11y) for bulk testing that reports any aXe accessibility violations from a set of pages. To use this script, a configuration file can be set to adapt our script to your specific needs like authentication, waiting for any specific items to finish loading (like a loading spinner), etc. 
```

 
 When used as a report, it will give an output [similar to this example](http://a11y-os.surge.sh/). The UI should be built first in the CI workflow and then a job created to run the a11y script against that build. The a11y script assumes that there is a webserver running somewhere that is serving up the pages, i.e. in localhost:8000, that it can reach. If you want to test a step before deployment, you could follow the path of integrating axe with cypress. If you are contributing to PatternFly, refer to our [README.md](//github.com/patternfly/patternfly/blob/main/README.md#testing-for-accessibility) on how to run this tool.

> ### 3. Test keyboard accessibility

Author: ```suggestion
### Test keyboard accessibility
```

Because the keyboard is a key accessibility tool, it is necessary to ensure that the following requirements are met: 

  - All functionality is keyboard accessible.
  - Elements in the HTML and in the layout follow a logical order.
  - Elements with focus are clearly visible.

> ### 4. Test without styles

Author: ```suggestion
### Test without styles
```

Because screen readers cannot access style information, you should disable styles for your product and ensure that your information architecture is effective and that there are adequate text labels.

 The [WAVE browser extension from WebAIM](//wave.webaim.org/extension/) provides this feature if it isn't available in the browser you are using.

> ### 5. Test with a screen reader 

Author: ```suggestion
### Test with a screen reader 
```

You can use any screen reader that is available in your operating system. To test PatternFly, we target the following screen readers:
  - JAWS with Chrome, Windows ([JAWS keyboard shortcuts](//dequeuniversity.com/screenreaders/jaws-keyboard-shortcuts))
  - Voiceover with Safari, Mac ([Voiceover keyboard shortcuts](//dequeuniversity.com/screenreaders/voiceover-keyboard-shortcuts))

>   - NVDA with Firefox, Windows ([NVDA keyboard shortcuts](//dequeuniversity.com/screenreaders/nvda-keyboard-shortcuts))

Reviewer 2: Nit: I think it'd be worth linking to the respective sites for the screen readers here as well if we could.

> ### 6. Check color contrast

Author: ```suggestion
### Check color contrast
```

Colors should pass the following contrast checks, to ensure that users across the vision spectrum can understand your product:
    - Text color against background color ([Understanding WCAG 1.4.3](//www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html))
    - Text color against link color ([Technique G183](//www.w3.org/TR/WCAG20-TECHS/G183.html))
    - Visible boundaries of buttons and form elements against adjacent background color ([Understanding WCAG 1.4.11](//www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html))

> 

Reviewer: I noticed that we removed the section on screen reader specific considerations. Do you think the information was redundant or not helpful? Perhaps we could take some of this information and put it underneath the screen reader section above to retain some of it? Curious of your thoughts on this one.

Reviewer: Realizing as I continued reviewing that this was added to the development section. I have mixed feelings. I remember adding it so that users understood how things differed in a screen reader for testing, but I totally get that this also affects developing. @thatblindgeye what do you think? I'm worried that being closer to the content/writing it, I might not be able to see it as objectively. 😂

Reviewer 2: Yeah that's tough because it can probably fit in either place. I feel like the content as it is *probably* fits closer in the development page. I guess it also depends if someone would be going through these pages in order or just what page interests them. I may have a clearer idea once I finish going through the current PRs but wondering whether linking to the development page would make sense? Sort of to convey "Also make sure you're testing stuff we mentioned on this page" or something, if that makes sense.

Reviewer: Yeah, that would make sense to me! I think after reviewing them a second time, I agree that that screen reader content probably makes more sense in the development section.

## Accessibility testing checklist

We recommend referencing the following checklist as you assess your product's accessibility.

> These are some of the main areas we check for on the PatternFly team to ensure that components meet consistent accessibility standards, but, to check your complete implementation, we recommend checking these same areas in your products. 

Reviewer: I'm wondering if we can tweak this sentence a little? Looking at the previous version, that probably needed tweaked as well in terms of what was meant. The PF team uses these criteria when checking a UI for a11y, not necessarily components, since some of these don't really apply to the component level. So for example, the entire section on structure is more of a UI consideration (at least in the way we're talking about it). Having heading structure of a page done correctly doesn't really apply to the component level, but when you're creating a page and potentially using multiple components.

### General criteria

> <label><input type="checkbox" /> VO Rotor Navigation can discover all information. </label><br/>
> <label><input type="checkbox" /> VO Shortcut Navigation can discover all information. </label><br/>
> <label><input type="checkbox" /> VO Cursor Navigation can discover all information (dependent on component and use case — there should be an easy way to navigate to the component). </label><br/>
> <label><input type="checkbox" /> Tabbed Keyboard Navigation can discover all information. </label>
> - Unless content should be hidden from a screen reader (such as offscreen or presentational information) make sure that content is set to aria-hidden=”true”.  <br/>
> 
> <label><input type="checkbox" /> Items are understandable and usable. </label><br/>
>   - <label><input type="checkbox" /> When you navigate to an item by keyboard or by screen reader, you should be able to easily understand and use the item. </label><br/>
>   - <label><input type="checkbox" /> The flow of information makes sense when navigating. Screen readers navigate the page in DOM order. If you used CSS to visually reposition elements, they might be announced in a nonsensical sequence. Instead, if you need something to appear earlier in the page, try to physically move it earlier in the DOM.</label>

Reviewer 2: So I only have an example of a change for this section of checklist items, but the same updates could be made to the next section as well:

```suggestion
<Checkbox label="Rotor navigation can discover all information, if a rotor menu exists." id="general-criteria-1" />
<Checkbox label="Shortcut navigation can discover all applicable information." description="For example, a keyboard shortcut that navigates between all headings on a page should discover all intended headings." id="general-criteria-2" />
<Checkbox label="Navigation via the assistive technology can discover all applicable information." description="Some assistive technologies will have their own means of navigation and focus management. For example, VoiceOver uses a 'VO' key plus right or left arrow keys to navigate a page. This may be different than conventional keyboard navigation." id="general-criteria-3" />
<Checkbox label={<span>Keyboard navigation via the <kbd>Tab</kbd> key can discover all information.</span>} description={<span>If content should be hidden from other assistive technology like a screen reader, <span className="ws-code">aria-hidden="true"</span> should be passed to it.</span>} id="general-criteria-4" />
<Checkbox label="Items are understandable and usable." body={<><Checkbox label="When you navigate to an item by keyboard or other assistive technology, you should be able to easily understand and use the item." id="general-criteria-5a" /><Checkbox label="The flow of information makes sense when navigating." description="Assistive technology like screen readers navigate the page in DOM order. If you used CSS to visually reposition elements, they might be announced in a nonsensical sequence. Instead, if you need something to appear earlier in the page, try to physically move it earlier in the DOM.
" id="general-criteria-5b" /></>} id="general-criteria-5" />
```

Some verbiage tweaks, but also using the PF checkbox component instead of the native HTML one. It requires a little more setup to use, especially when it comes to using other HTML tags and links, but it results in the following:

![General criteria checklist with PatternFly Checkbox](https://github.com/patternfly/patternfly-org/assets/70952936/f4539f19-fecb-4430-8990-d4ea19636a45)

Author: for your suggestions on line 73 & 74 - it looks like some characters aren't allowed within <Checkbox>, like code snippets & "body={<>" - at least for me. Did you create your screenshot via vs code, without any errors?

> ### Accessibility points

Reviewer 2: I'm not really sure what they could be updated to, but I feel like one or both of the headings for these sections of checklists could be updated from "General criteria" and "Accessibility points". Maybe just this section, as the entire page is regarding accessibility points technically?

Also not sure whether the current/previous content here still fits or if it's either duplicate from the "general criteria" or needs a little more information. The "Links" item in the Structure section, for example, I'm not sure of the purpose there since we mention them again in the Labels section. @jessiehuff wdyt?

Author: I agree on this! I can't immediately tell the difference between the lists, which pushes me towards a heading change or combining the two lists. But I'm not entirely sure what the common thread between the "accessibility points" section is, so I'm not sure how to go about renaming

Reviewer: Definitely open to different names! My original intention was that the first section was thinking about the different ways of checking things (i.e. screen reader and it's different ways of navigating and keyboard) and that essentially everything should work through each of the general a11y questions and thinking big picture (i.e. just like the questions asked in the development section - is it discoverable, able to be interacted with, and understandable). The second section was more getting into the nitty gritty of checking specific things, starting from high level like structure to more granular things like labels. So totally open to anything that communicates that better! If both of you were confused, it definitely sounds like something should be changed!

In terms of the links in both the structure and labels section because when I'm testing structure, especially using the rotor menu, I check each of those major elements (headings, landmarks, links, form controls, tables) and make sure that nothing immediately looks off with the way that they're handled. That's probably not totally clear though. 🙃 I guess that's what happens when a checklist starts off as more of something for your personal use to trying to communicate it more publicly haha.

section: accessibility
---

# Testing your product's accessibility 

This guide contains instructions and recommendations to comprehensively test your product's accessibility to identify accessibility issues and opportunities for improvement.  

> **Keep in mind that this guide may not cover every scenario.**
> 

Author: ```suggestion
**Keep in mind that this guide may not cover every scenario.**
```

## General testing recommendations 

Many accessibility issues can be found by doing a few simple checks: 

### Validate your HTML

Good accessibility practices start with structural, semantic HTML. When a screen reader (or any sort of assistive technology) scans a web page, it gets information about the Document Object Model (DOM), or the HTML structure of the page. No styles or JavaScript will be read by a screen reader. 

> Screen reader software like Voice Over (VO), NVDA, or JAWS doesn’t just turn text into speech. It can use information in the HTML to list all of the headings on a page, give extra navigation controls to data tables, announce how many items are in a list, and more. This makes semantic HTML essential. 

Author: ```suggestion
Screen readers like Voice Over (VO), NVDA, or JAWS don't just turn text into speech. They use information in the HTML to list all of the headings on a page, give extra navigation controls to data tables, announce how many items are in a list, and more. This makes semantic HTML essential. 
```

There are many tools you can use to validate your HTML, such as [W3C’s markup validation service](https://validator.w3.org/). 

### Check for accessibility violations with an audit tool

 If you are using PatternFly, we recommend using aXe: The Accessibility Engine (using their [DevTools](https://www.deque.com/axe/devtools/) or the [Chrome](https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd) or [Firefox](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/) extension) to check for accessibility violations locally. 
 
We also offer the [patternfly-a11y script](https://github.com/patternfly/patternfly-a11y) for bulk testing that reports any aXe accessibility violations from a set of pages. To use this script, a configuration file can be set to adapt our script to your specific needs like authentication, waiting for any specific items to finish loading (like a loading spinner), etc. 
 

>  When used as a report, it will give an [accessibility report via surge](http://a11y-os.surge.sh/). The UI should be built first in the CI workflow and then a job created to run the a11y script against that build. The a11y script assumes that there is a webserver running somewhere that is serving up the pages, i.e. in localhost:8000, that it can reach. If you want to test a step before deployment, you could follow the path of integrating axe with cypress.

Author: ```suggestion
 When used as a report, it will give an [accessibility report via surge](http://a11y-os.surge.sh/). The UI should be built first in the CI workflow and then a job created to run the a11y script against that build. The a11y script assumes that there is a web server running somewhere that is serving up the pages  that it can reach. (for example, in localhost:8000). If you want to test a step before deployment, you could follow the path of integrating aXe with [Cypress](https://www.cypress.io/).
```

### Test keyboard accessibility

> Because the keyboard is a key accessibility tool, it is necessary to ensure that the following requirements are met: 

Author: ```suggestion
The keyboard is an essential accessibility tool, so it is necessary to ensure that the following requirements are met: 
```
